pkg/distributor/ha_tracker.go

```diff
@@ -202,10 +202,55 @@ func newHATracker(cfg HATrackerConfig, limits haTrackerLimits, reg prometheus.Re
 		t.client = client
 	}
 
-	t.Service = services.NewBasicService(nil, t.loop, nil)
+	t.Service = services.NewBasicService(t.syncHAStateOnStart, t.loop, nil)
 	return t, nil
 }
 
+// Will update the internal cache to be in sync with the KV Store
+func (h *haTracker) syncHAStateOnStart(ctx context.Context) error {
+	if !h.cfg.EnableHATracker {
+		return nil
+	}
+
+	// haTracker holds HATrackerConfig with the prefix.
+	keys, err := h.client.List(ctx, "")
+	if err != nil {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to list replica keys", "err", err)
+		return err
+	}
+
+	if len(keys) == 0 {
+		level.Warn(h.logger).Log("msg", "syncHAStateOnStart: no keys for HA tracker prefix", "err", err)
+		return nil
+	}
+
+	// TODO what is the benefit to allowing an a concurrent/parallel process here
+	// What is the amount fo keys we are getting back? Is it inefficient to serially allow this behavior?
+	for i := 0; i < len(keys); i++ {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		val, err := h.client.Get(ctx, keys[i])
+		if err != nil {
+			level.Warn(h.logger).Log("msg", "syncHAStateOnStart: failed to get replica value", "key", keys[i], "err", err)
+			return err
```

Comment:
There were certain errors  I thought that matters to return out. But you are right a lot of there i should continue just logging if there is an error. The only error that makes sense to return out would be the one where if the Context was canceled so something bigger happened in the system so we should return out there, that sound okay?

Reply:
> 1. What if for the `h.client` there were bigger issue's (like a connectivity issues) which made me think the `startingFn` behavior would have issues since it uses the `h.client`

We could potentially use retries with backoff in case the KV store has issues. 

I just checked the regular HA tracker logic and changed my mind. On the regular push path we return a 5xx error in case the KV store is unavailable. I think on startup it makes more sense to follow the same logic. That is - instead of `continue`ing we return an error. We should still stick to the error tolerance that the regular checkSample has (e.g. if the key doesn't follow the expected 2-segmented format).

WDYT?